File: src/main/java/xyz/phanta/cbtweaker/structure/MatcherCuboid.java
package xyz.phanta.cbtweaker.structure;

import gnu.trove.map.TCharObjectMap;
import net.minecraft.util.math.BlockPos;
import net.minecraft.util.math.Vec3i;
import net.minecraft.world.World;
import org.apache.commons.lang3.tuple.Pair;
import xyz.phanta.cbtweaker.structure.block.StructureBlockMatch;
import xyz.phanta.cbtweaker.structure.block.StructureBlockMatcher;
import xyz.phanta.cbtweaker.util.ConfigException;
import xyz.phanta.cbtweaker.util.Direction;

import javax.annotation.Nullable;
import java.util.Collection;
import java.util.HashMap;
import java.util.Map;
import java.util.stream.Stream;

public class MatcherCuboid {

    private final Map<Vec3i, Pair<Vec3i, StructureBlockMatcher>> posMatcherTable = new HashMap<>();
    private final int minX, maxX, minY, maxY, minZ, maxZ;

    public MatcherCuboid(char[][][] blockArray, Vec3i origin, TCharObjectMap<StructureBlockMatcher> palette) {
        int minX = Integer.MAX_VALUE, maxX = Integer.MIN_VALUE;
        int minY = Integer.MAX_VALUE, maxY = Integer.MIN_VALUE;
        int minZ = Integer.MAX_VALUE, maxZ = Integer.MIN_VALUE;
        for (int arrY = 0; arrY < blockArray.length; arrY++) {
            char[][] ySlice = blockArray[arrY];
            for (int arrZ = 0; arrZ < ySlice.length; arrZ++) {
                char[] zSlice = ySlice[arrZ];
                for (int arrX = 0; arrX < zSlice.length; arrX++) {
                    char matcherChar = zSlice[arrX];
                    if (matcherChar == ' ') {
                        continue;
                    }
                    StructureBlockMatcher matcher = palette.get(matcherChar);
                    if (matcher == null) {
                        throw new ConfigException("Character not in palette: " + matcherChar);
                    }
                    int x = arrX - origin.getX(), y = arrY - origin.getY(), z = arrZ - origin.getZ();
                    if (x < minX) {
                        minX = x;
                    }
                    if (x > maxX) {
                        maxX = x;
                    }
                    if (y < minY) {
                        minY = y;
                    }
                    if (y > maxY) {
                        maxY = y;
                    }
                    if (z < minZ) {
                        minZ = z;
                    }
                    if (z > maxZ) {
                        maxZ = z;
                    }
                    Vec3i offset = new Vec3i(x, y, z);
                    posMatcherTable.put(offset, Pair.of(offset, matcher));
                }
            }
        }
        this.minX = minX;
        this.maxX = maxX;
        this.minY = minY;
        this.maxY = maxY;
        this.minZ = minZ;
        this.maxZ = maxZ;
    }

    public Collection<Pair<Vec3i, StructureBlockMatcher>> getMatchers() {
        return posMatcherTable.values();
    }

    public Stream<BlockPos> computePositions(BlockPos originPos, Direction dir, boolean mirror) {
        return posMatcherTable.values().stream().map(s -> computeOffsetPos(originPos, s.getLeft(), dir, mirror));
    }

    public int getMinX() {
        return minX;
    }

    public int getMaxX() {
        return maxX;
    }

    public int getSizeX() {
        return maxX - minX + 1;
    }

    public int getMinY() {
        return minY;
    }

    public int getMaxY() {
        return maxY;
    }

    public int getSizeY() {
        return maxY - minY + 1;
    }

    public int getMinZ() {
        return minZ;
    }

    public int getMaxZ() {
        return maxZ;
    }

    public int getSizeZ() {
        return maxZ - minZ + 1;
    }

    @Nullable
    public StructureMatch tryMatch(World world, BlockPos originPos, Direction dir, boolean mirror) {
        StructureMatch structMatch = new StructureMatch();
        for (Pair<Vec3i, StructureBlockMatcher> posMatcher : posMatcherTable.values()) {
            BlockPos pos = computeOffsetPos(originPos, posMatcher.getLeft(), dir, mirror);
            StructureBlockMatch blockMatch = posMatcher.getRight().matchBlock(world, pos, dir);
            if (blockMatch == null) {
                return null;
            }
            structMatch.addPosition(pos);
            if (blockMatch instanceof StructureBlockMatch.Component) {
                StructureBlockMatch.Component compMatch = (StructureBlockMatch.Component)blockMatch;
                structMatch.addComponent(compMatch.getComponentId());
            } else if (blockMatch instanceof StructureBlockMatch.Hatch) {
                StructureBlockMatch.Hatch hatchMatch = (StructureBlockMatch.Hatch)blockMatch;
                structMatch.addHatch(hatchMatch.getGroupId(), hatchMatch.getHatch());
            }
        }
        return structMatch;
    }

    private static BlockPos computeOffsetPos(BlockPos basePos, Vec3i offset, Direction dir, boolean mirror) {
        return basePos.add(dir.transform(offset, mirror));
    }

}

File: src/main/kotlin/st/evening/mc/cbtweaker/structure/MatcherCuboid.kt
package st.evening.mc.cbtweaker.structure

import it.unimi.dsi.fastutil.chars.Char2ObjectMap
import net.minecraft.util.Rotation
import net.minecraft.util.math.BlockPos
import net.minecraft.util.math.Vec3i
import net.minecraft.world.World
import st.evening.mc.cbtweaker.structure.block.StructureBlockMatch
import st.evening.mc.cbtweaker.structure.block.StructureBlockMatcher
import st.evening.mc.cbtweaker.util.invOffsetWithRotation
import st.evening.mc.cbtweaker.util.offsetWithRotation
import st.evening.mc.prelude.api.data.ser.SerializationException
import st.evening.mc.prelude.api.data.tjson.JsonPath

class MatcherCuboid private constructor(
    val posMatcherTable: Map<Vec3i, StructureBlockMatcher>,
    val minX: Int,
    val maxX: Int,
    val minY: Int,
    val maxY: Int,
    val minZ: Int,
    val maxZ: Int
) {
    companion object {
        context(_: JsonPath)
        fun load(
            blueprint: Array<Array<CharArray>>,
            origin: Vec3i,
            palette: Char2ObjectMap<StructureBlockMatcher>
        ): MatcherCuboid {
            val posMatcherTable = mutableMapOf<Vec3i, StructureBlockMatcher>()
            val originX = origin.x
            val originY = origin.y
            val originZ = origin.z
            var minX = Int.MAX_VALUE
            var maxX = Int.MIN_VALUE
            var minY = Int.MAX_VALUE
            var maxY = Int.MIN_VALUE
            var minZ = Int.MAX_VALUE
            var maxZ = Int.MIN_VALUE
            for (arrY in blueprint.indices) {
                val ySlice = blueprint[arrY]
                for (arrZ in ySlice.indices) {
                    val zSlice = ySlice[arrZ]
                    for (arrX in zSlice.indices) {
                        val c = zSlice[arrX]
                        if (c == ' ') continue
                        val matcher = palette[c]
                            ?: throw SerializationException.withPath("Character not in palette: $c")
                        val x = arrX - originX
                        val y = arrY - originY
                        val z = arrZ - originZ
                        if (x < minX) minX = x
                        if (x > maxX) maxX = x
                        if (y < minY) minY = y
                        if (y > maxY) maxY = y
                        if (z < minZ) minZ = z
                        if (z > maxZ) maxZ = z
                        posMatcherTable[Vec3i(x, y, z)] = matcher
                    }
                }
            }
            return MatcherCuboid(posMatcherTable, minX, maxX, minY, maxY, minZ, maxZ)
        }
    }

    val sizeX: Int
        get() = maxX - minX + 1
    val sizeY: Int
        get() = maxY - minY + 1
    val sizeZ: Int
        get() = maxZ - minZ + 1

    fun computePositions(originPos: BlockPos, rotation: Rotation, mirrorX: Boolean): Sequence<BlockPos> =
        originPos.offsetWithRotation(posMatcherTable.keys.asSequence(), rotation, mirrorX)

    fun tryMatch(world: World, originPos: BlockPos, rotation: Rotation, mirrorX: Boolean): List<StructureBlockMatch>? {
        return posMatcherTable.map { (offset, matcher) ->
            matcher.matchBlock(world, originPos.offsetWithRotation(offset, rotation, mirrorX), rotation, mirrorX)
                ?: return null
        }
    }

    class IncrementalMatcher(
        val matchRegion: MatcherCuboid,
        val world: World,
        val originPos: BlockPos,
        val rotation: Rotation,
        val mirrorX: Boolean
    ) {
        private val toCheck: MutableSet<BlockPos> =
            matchRegion.computePositions(originPos, rotation, mirrorX).toMutableSet()
        private val failedChecks: MutableSet<BlockPos> = mutableSetOf()
        private val cachedMatch: MutableMap<BlockPos, StructureBlockMatch> = mutableMapOf()

        var lastMatchSuccess: Boolean? = null
            private set

        val dirtyCount: Int
            get() = toCheck.size + failedChecks.size

        // returns garbage if lastMatchSuccess != true
        fun getLastMatch(): Map<BlockPos, StructureBlockMatch> = cachedMatch

        fun markDirty(pos: BlockPos) {
            toCheck += pos
            failedChecks -= pos
            cachedMatch -= pos
            lastMatchSuccess = null
        }

        fun markDirty(positions: Collection<BlockPos>) {
            toCheck.addAll(positions)
            val positionSet = positions.toSet()
            failedChecks.removeAll(positionSet)
            cachedMatch -= positionSet
            lastMatchSuccess = null
        }

        fun tryMatch(): Map<BlockPos, StructureBlockMatch>? {
            toCheck.forEach { pos ->
                val matcher = matchRegion.posMatcherTable[pos.invOffsetWithRotation(originPos, rotation, mirrorX)]
                    ?: return@forEach
                val match = matcher.matchBlock(world, pos, rotation, mirrorX)
                if (match != null) {
                    cachedMatch[pos] = match
                } else {
                    failedChecks += pos
                }
            }
            toCheck.clear()
            if (failedChecks.isEmpty()) {
                lastMatchSuccess = true
                // could copy this, but it's unlikely that the old result will be needed after a new tryMatch call
                return cachedMatch
            } else {
                lastMatchSuccess = false
                return null
            }
        }

        fun computePositions(): Sequence<BlockPos> = matchRegion.computePositions(originPos, rotation, mirrorX)
    }
}
